bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/obsrecord/VisitCalculator.scala

```diff
@@ -167,8 +183,112 @@ private[obsrecord] object VisitCalculator {
 
   }
 
+  def charges(
+    events:     VisitEvents,
+    instrument: Option[Instrument],
+    obsClass:   ObsClass,
+    datasetQa:  DatasetLabel => DatasetQaState,
+    datasetOc:  DatasetLabel => ObsClass
+  ): IntervalCharges =
+    new IntervalCharges(events, instrument, obsClass, datasetQa, datasetOc)
+
+  /**
+   * The 2019A calculator.  The main change from the previous time accounting
+   * result is that visits without a passing dataset are not charged at all,
+   * except for visitor instrument observations which are charged regardless.
+   * It also simplifies and fixes a number of bugs with the original algorithm:
+   *
+   * <ul>
+   *   <li>Charges for the portion of a dataset after start and before overlap</li>
+   *   <li>Handles start/end dataset pairs with intervening events correctly</li>
+   *   <li>Doesn't drop time if the first event is start dataset</li>
+   * </ul>
+   */
+  case object Update2019A extends SemesterVisitCalculator(new Semester(2019, Semester.Half.A)) {
+
+    override def calc(
+      events:     VisitEvents,
+      instrument: Option[Instrument],
+      obsClass:   ObsClass,
+      datasetQa:  DatasetLabel => DatasetQaState,
+      datasetOc:  DatasetLabel => ObsClass,
+      nextVisit:  Option[(VisitTimes, VisitEvents)] // ignored here
+    ): VisitTimes =
+      charges(events, instrument, obsClass, datasetQa, datasetOc).when { c =>
+        c.isVisitor || c.hasChargeableDataset || c.isGpiAcquisition
+      }
+
+  }
+
+  /**
+   * Updates the 2019A visit calculator to charge for slew visits.  The typical
+   * pattern is to slew on the science observation, switch to a special
+   * acquisition observation, then switch back to the science observation to
+   * collect data.  The initial slew was not being charged.
+   */
+  case object Update2019B extends SemesterVisitCalculator(new Semester(2019, Semester.Half.B)) {
+
+    override def calc(
+      events:     VisitEvents,
+      instrument: Option[Instrument],
+      obsClass:   ObsClass,
+      datasetQa:  DatasetLabel => DatasetQaState,
+      datasetOc:  DatasetLabel => ObsClass,
+      nextVisit:  Option[(VisitTimes, VisitEvents)]
+    ): VisitTimes =
+      charges(events, instrument, obsClass, datasetQa, datasetOc).when { c =>
+        c.isVisitor              ||
+          c.hasChargeableDataset ||
+          c.isGpiAcquisition     ||
+          c.isChargeableSlew(nextVisit)
+      }
+  }
+
   // reverse order by valid time (newest first)
   def all: List[VisitCalculator] =
-    List(Update2019A, Primordial)
+    List(Update2019B, Update2019A, Primordial)
+
+  /**
+   * Finds the `VisitCalculator` that corresponds to the given visit.
+   */
+  def lookup(e: VisitEvents): VisitCalculator =
+    (for {
+      h <- e.sorted.headOption
+      s <- h.site
+      c <- all.find(_.validAt(s).isBefore(h.instant))
+    } yield c).getOrElse(all.head)
```

Comment:
Would it be a hassle to make `all` a NEL while we're mucking around in here?

Reply:
Good point. It's only used in one place so it was easy to switch.